File: ios/Classes/boundings/map/port/MapApi.swift
import Flutter
import Foundation
import UIKit
import MapboxMaps
import MapboxDirections
import MapboxCoreNavigation
import MapboxNavigation

public class MapAPI: NSObject, FlutterStreamHandler 
{
    var _eventSink: FlutterEventSink? = nil
    private var mapboxMap: MapboxMap

    let messenger: FlutterBinaryMessenger
    let channel: FlutterMethodChannel
    let eventChannel: FlutterEventChannel

    init(messenger: FlutterBinaryMessenger, withMapboxMap mapboxMap: MapboxMap, viewId: Int64) {
        self.mapboxMap = mapboxMap
        
        self.messenger = messenger
        self.channel = FlutterMethodChannel(name: "flutter_mapbox_navigation/map/\(viewId)", binaryMessenger: messenger, codec: FLT_MapInterfaceGetCodec())
        self.eventChannel = FlutterEventChannel(name: "flutter_mapbox_navigation/map/\(viewId)/events", binaryMessenger: messenger, codec: FLT_MapInterfaceGetCodec())

        super.init()

        self.eventChannel.setStreamHandler(self)

        self.channel.setMethodCallHandler { [weak self](call, result) in

            guard let strongSelf = self else { return }

            let arguments = call.arguments as? NSDictionary

            if(call.method == "pixelForCoordinate")
            {
                strongSelf.pixelForCoordinate(arguments: arguments, result: result)
            } 
            else if (call.method == "queryRenderedFeatures") 
            {
                strongSelf.queryRenderedFeatures(arguments: arguments, result: result)
            }
            else
            {
                result("method is not implemented");
            }
        }

        subscribeEvents()
    }

    func pixelForCoordinate(arguments: NSDictionary?, result: @escaping FlutterResult) {
        do {
            guard let lat = arguments?["latitude"] as? Double else {return}
            guard let lon = arguments?["longitude"] as? Double else {return}
            let coordinate = CLLocationCoordinate2D(latitude: lat, longitude: lon)
            let point = self.mapboxMap.point(for: coordinate)
            result(point.toFLTScreenCoordinate())
        } catch {
            result(FlutterError(code: MapAPI.errorCode, message: "\(error)", details: nil))
        }
    }

    func queryRenderedFeatures(arguments: NSDictionary?, result: @escaping FlutterResult) {
        guard let geometry = arguments?["geometry"] as? FLTRenderedQueryGeometry else {return}
        guard let options = arguments?["options"] as? FLTRenderedQueryOptions else {return}
        do {
            if geometry.type == FLTType.SCREEN_BOX {
                let screenBoxArray = convertStringToArray(properties: geometry.value)
                guard let minCoord = screenBoxArray[0] as? [Double] else {return}
                guard let maxCoord = screenBoxArray[1] as? [Double] else {return}

                let screenBox = ScreenBox(min: ScreenCoordinate(x: minCoord[0], y: minCoord[1]),
                                          max: ScreenCoordinate(x: maxCoord[0], y: maxCoord[1]))
                let cgRect = screenBox.toCGRect()
                let queryOptions = try options.toRenderedQueryOptions()
                self.mapboxMap.queryRenderedFeatures(in: cgRect, options: queryOptions) { res in
                    switch res {
                    case .success(let features):
                        result(features.map({$0.toFLTQueriedFeature()}))
                    case .failure(let error):
                        result(FlutterError(code: MapAPI.errorCode, message: "\(error)", details: nil))
                    }
                }
            } else if geometry.type == FLTType.SCREEN_COORDINATE {
                guard let pointArray = convertStringToArray(properties: geometry.value) as? [Double] else {return}
                let cgPoint = CGPoint(x: pointArray[0], y: pointArray[1])

                try self.mapboxMap.queryRenderedFeatures(at: cgPoint, options: options.toRenderedQueryOptions()) { res in
                    switch res {
                    case .success(let features):
                        result(features.map({$0.toFLTQueriedFeature()}))
                    case .failure(let error):
                        result(FlutterError(code: MapAPI.errorCode, message: "\(error)", details: nil))
                    }
                }
            } else {
                let cgPoints = try JSONDecoder().decode([[Double]].self, from: geometry.value.data(using: String.Encoding.utf8)!)

                try self.mapboxMap.queryRenderedFeatures(for: cgPoints.map({CGPoint(x: $0[0], y: $0[1])}), options: options.toRenderedQueryOptions()) { res in
                    switch res {
                    case .success(let features):
                        result(features.map({$0.toFLTQueriedFeature()}))
                    case .failure(let error):
                        result(FlutterError(code: MapAPI.errorCode, message: "\(error)", details: nil))
                    }
                }
            }
        } catch {
            result(FlutterError(code: MapAPI.errorCode, message: "\(error)", details: nil))
        }
    }

    private func subscribeEvents() {
        self.mapboxMap.onEvery(event: MapEvents.Event.cameraChanged!) { (event) in
            guard let data = event.payload as? [String: Any] else {return}
            self.channel.invokeMethod(self.getEventMethodName(eventType: event.name),
                                        arguments: self.convertDictionaryToString(dict: data))
        }

        self.mapboxMap.onEvery(event: MapEvents.Event.mapIdle!) { (event) in
            guard let data = event.payload as? [String: Any] else {return}
            self.channel.invokeMethod(self.getEventMethodName(eventType: event.name),
                                        arguments: self.convertDictionaryToString(dict: data))
        }

        self.mapboxMap.onEvery(event: MapEvents.Event.mapLoaded!) { (event) in
            guard let data = event.payload as? [String: Any] else {return}
            self.channel.invokeMethod(self.getEventMethodName(eventType: event.name),
                                        arguments: self.convertDictionaryToString(dict: data))
        }

        self.mapboxMap.onEvery(event: MapEvents.Event.renderFrameFinished!) { (event) in
            guard let data = event.payload as? [String: Any] else {return}
            self.channel.invokeMethod(self.getEventMethodName(eventType: event.name),
                                        arguments: self.convertDictionaryToString(dict: data))
        }

        self.mapboxMap.onEvery(event: MapEvents.Event.renderFrameStarted!) { (event) in
            guard let data = event.payload as? [String: Any] else {return}
            self.channel.invokeMethod(self.getEventMethodName(eventType: event.name),
                                        arguments: self.convertDictionaryToString(dict: data))
        }

        self.mapboxMap.onEvery(event: MapEvents.Event.sourceAdded!) { (event) in
            guard let data = event.payload as? [String: Any] else {return}
            self.channel.invokeMethod(self.getEventMethodName(eventType: event.name),
                                        arguments: self.convertDictionaryToString(dict: data))
        }

        self.mapboxMap.onEvery(event: MapEvents.Event.sourceDataLoaded!) { (event) in
            guard let data = event.payload as? [String: Any] else {return}
            self.channel.invokeMethod(self.getEventMethodName(eventType: event.name),
                                        arguments: self.convertDictionaryToString(dict: data))
        }

        self.mapboxMap.onEvery(event: MapEvents.Event.sourceRemoved!) { (event) in
            guard let data = event.payload as? [String: Any] else {return}
            self.channel.invokeMethod(self.getEventMethodName(eventType: event.name),
                                        arguments: self.convertDictionaryToString(dict: data))
        }

        self.mapboxMap.onEvery(event: MapEvents.Event.styleDataLoaded!) { (event) in
            guard let data = event.payload as? [String: Any] else {return}
            self.channel.invokeMethod(self.getEventMethodName(eventType: event.name),
                                        arguments: self.convertDictionaryToString(dict: data))
        }

        self.mapboxMap.onEvery(event: MapEvents.Event.styleImageMissing!) { (event) in
            guard let data = event.payload as? [String: Any] else {return}
            self.channel.invokeMethod(self.getEventMethodName(eventType: event.name),
                                        arguments: self.convertDictionaryToString(dict: data))
        }

        self.mapboxMap.onEvery(event: MapEvents.Event.styleImageRemoveUnused!) { (event) in
            guard let data = event.payload as? [String: Any] else {return}
            self.channel.invokeMethod(self.getEventMethodName(eventType: event.name),
                                        arguments: self.convertDictionaryToString(dict: data))
        }

        self.mapboxMap.onEvery(event: MapEvents.Event.styleLoaded!) { (event) in
            guard let data = event.payload as? [String: Any] else {return}
            self.channel.invokeMethod(self.getEventMethodName(eventType: event.name),
                                        arguments: self.convertDictionaryToString(dict: data))
        }
    }

    private func getEventMethodName(eventType: String) -> String {
        return "event#\(eventType)"
    }

    private func convertDictionaryToString(dict: [String: Any]) -> String {
        var result: String = ""
        do {
            let jsonData =
            try JSONSerialization.data(
                withJSONObject: dict,
                options: JSONSerialization.WritingOptions.init(rawValue: 0)
            )

            if let JSONString = String(data: jsonData, encoding: String.Encoding.utf8) {
                result = JSONString
            }
        } catch {
            result = ""
        }
        return result
    }
    
    //MARK: EventListener Delegates
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        _eventSink = events
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        _eventSink = nil
        return nil
    }

    private static let errorCode = "map"
}

File: android/src/main/kotlin/com/eopeter/fluttermapboxnavigation/boundings/map/port/MapApi.kt
package com.eopeter.fluttermapboxnavigation.boundings.map.port

import android.content.Context
import com.mapbox.maps.extension.observable.*
import com.eopeter.fluttermapboxnavigation.boundings.map.application.MapApiCodec
import com.eopeter.fluttermapboxnavigation.core.*
import com.google.gson.GsonBuilder
import com.mapbox.geojson.Point
import com.mapbox.maps.MapboxMap
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.StandardMethodCodec

class MapApi : MethodChannel.MethodCallHandler {
    private var methodChannel: MethodChannel? = null
    private val messenger: BinaryMessenger
    private val mapboxMap: MapboxMap
    private val context: Context
    private val viewId: Int
    private val gson = GsonBuilder().create()

    constructor(messenger: BinaryMessenger, mapboxMap: MapboxMap, viewId: Int, context: Context) {
        this@MapApi.messenger = messenger
        this@MapApi.mapboxMap = mapboxMap
        this@MapApi.viewId = viewId
        this@MapApi.context = context
    }

    fun init() {
        val methodChannel =
            MethodChannel(
                this.messenger,
                "flutter_mapbox_navigation/map/${this.viewId}",
                StandardMethodCodec(MapApiCodec)
            )
        methodChannel.setMethodCallHandler(this)
        this.listenOnEvents(methodChannel)
        this.methodChannel = methodChannel
    }

    override fun onMethodCall(methodCall: MethodCall, result: MethodChannel.Result) {
        when (methodCall.method) {
            "pixelForCoordinate" -> {
                this.pixelForCoordinate(methodCall, result)
            }
            "queryRenderedFeatures" -> {
                this.queryRenderedFeatures(methodCall, result)
            }
        }
    }

    private fun queryRenderedFeatures(methodCall: MethodCall, result: MethodChannel.Result) {
        val arguments = methodCall.arguments as? Map<*, *> ?: return
        val geometry = arguments["geometry"] as? RenderedQueryGeometry ?: return
        val options = arguments["options"] as? RenderedQueryOptions ?: return
        mapboxMap.queryRenderedFeatures(
            geometry.toRenderedQueryGeometry(context),
            options.toRenderedQueryOptions()
        ) {
            if (it.isError) {
                result.success(it.error)
            } else {
                result.success(it.value!!.map { feature -> feature.toFLTQueriedFeature() }.toMutableList())
            }
        }
    }

    private fun pixelForCoordinate(methodCall: MethodCall, result: MethodChannel.Result) {
        val arguments = methodCall.arguments as? Map<*, *> ?: return
        val latitude = arguments["latitude"] as? Double ?: return
        val longitude = arguments["longitude"] as? Double ?: return
        val point = Point.fromLngLat(longitude, latitude)
        val screenCoordinate = mapboxMap.pixelForCoordinate(point)
        result.success(screenCoordinate.toFLTScreenCoordinate(context))
    }

    private fun listenOnEvents(methodChannel: MethodChannel) {
        mapboxMap.subscribeMapLoaded {methodChannel.invokeMethod(MapEvent.MAP_LOADED.methodName, gson.toJson(it))}
        mapboxMap.subscribeMapLoadingError {methodChannel.invokeMethod(MapEvent.MAP_LOADING_ERROR.methodName, gson.toJson(it))}
        mapboxMap.subscribeStyleLoaded {methodChannel.invokeMethod(MapEvent.STYLE_LOADED.methodName, gson.toJson(it))}
        mapboxMap.subscribeStyleDataLoaded {methodChannel.invokeMethod(MapEvent.STYLE_DATA_LOADED.methodName, gson.toJson(it))}
        mapboxMap.subscribeCameraChange {methodChannel.invokeMethod(MapEvent.CAMERA_CHANGED.methodName, gson.toJson(it))}
        mapboxMap.subscribeMapIdle {methodChannel.invokeMethod(MapEvent.MAP_IDLE.methodName, gson.toJson(it))}
        mapboxMap.subscribeSourceAdded {methodChannel.invokeMethod(MapEvent.SOURCE_ADDED.methodName, gson.toJson(it))}
        mapboxMap.subscribeSourceRemoved {methodChannel.invokeMethod(MapEvent.SOURCE_REMOVED.methodName, gson.toJson(it))}
        mapboxMap.subscribeSourceDataLoaded {methodChannel.invokeMethod(MapEvent.SOURCE_DATA_LOADED.methodName, gson.toJson(it))}
        mapboxMap.subscribeStyleImageMissing {methodChannel.invokeMethod(MapEvent.STYLE_IMAGE_MISSING.methodName, gson.toJson(it))}
        mapboxMap.subscribeStyleImageUnused {methodChannel.invokeMethod(MapEvent.STYLE_IMAGE_REMOVE_UNUSED.methodName, gson.toJson(it))}
        mapboxMap.subscribeRenderFrameStarted {methodChannel.invokeMethod(MapEvent.RENDER_FRAME_STARTED.methodName, gson.toJson(it))}
        mapboxMap.subscribeRenderFrameFinished {methodChannel.invokeMethod(MapEvent.RENDER_FRAME_FINISHED.methodName, gson.toJson(it))}
        mapboxMap.subscribeResourceRequest {methodChannel.invokeMethod(MapEvent.RESOURCE_REQUEST.methodName, gson.toJson(it))}
    }
}